nav2_costmap_2d/plugins/static_layer.cpp

```diff
@@ -56,232 +58,171 @@ using nav2_costmap_2d::FREE_SPACE;
-
-  std::string map_topic;
-  node_->get_parameter_or<std::string>("map_topic", map_topic, std::string("/map"));
-  node_->get_parameter_or<bool>("first_map_only", first_map_only_, false);
-  node_->get_parameter_or<bool>("subscribe_to_updates", subscribe_to_updates_, false);
-
-  node_->get_parameter_or<bool>("track_unknown_space", track_unknown_space_, true);
-  node_->get_parameter_or<bool>("use_maximum", use_maximum_, false);
-
-  int temp_lethal_threshold, temp_unknown_cost_value;
-  node_->get_parameter_or<int>("lethal_cost_threshold", temp_lethal_threshold, 100);
-  node_->get_parameter_or<int>("unknown_cost_value", temp_unknown_cost_value, -1);
-  node_->get_parameter_or<bool>("trinary_costmap", trinary_costmap_, true);
-
-  lethal_threshold_ = std::max(std::min(temp_lethal_threshold, 100), 0);
-  unknown_cost_value_ = temp_unknown_cost_value;
-
-  // we'll subscribe to the latched topic that the map server uses
-  RCLCPP_INFO(node_->get_logger(), "Requesting the map...");
-  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
-  custom_qos_profile.depth = 1;
-  custom_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
-  map_sub_ = node_->create_subscription<nav_msgs::msg::OccupancyGrid>(map_topic,
-      std::bind(&StaticLayer::incomingMap, this, std::placeholders::_1), custom_qos_profile);
-  map_received_ = false;
-  has_updated_data_ = false;
+}
 
-  rclcpp::Rate r(10);
-  rclcpp::executors::SingleThreadedExecutor exec;
-  while (!map_received_ && rclcpp::ok()) {
-    exec.spin_node_once(node_->get_node_base_interface(), std::chrono::milliseconds(100));
-    r.sleep();
-  }
+void
+StaticLayer::onInitialize()
+{
+  getParameters();
+  getMap();
 
-  RCLCPP_INFO(node_->get_logger(),
-    "Received a %d X %d map at %f m/pix", getSizeInCellsX(),
-    getSizeInCellsY(), getResolution());
+  current_ = true;
+}
 
-  if (subscribe_to_updates_) {
-    RCLCPP_INFO(node_->get_logger(), "Subscribing to updates");
-    map_update_sub_ = node_->create_subscription<map_msgs::msg::OccupancyGridUpdate>(
-      map_topic + "_updates",
-      std::bind(&StaticLayer::incomingUpdate, this, std::placeholders::_1), custom_qos_profile);
+void
+StaticLayer::activate()
+{
+}
 
-  } else {
-    has_updated_data_ = true;
-  }
+void
+StaticLayer::deactivate()
+{
+}
 
-  dynamic_param_client_ = std::make_unique<nav2_dynamic_params::DynamicParamsClient>(node_);
-  dynamic_param_client_->add_parameters({name_ + "." + "enabled"});
-  dynamic_param_client_->set_callback(std::bind(&StaticLayer::reconfigureCB, this));
-  // TODO(bpwilcox): Add new parameters to parameter validation class from plugins
+void
+StaticLayer::reset()
+{
+  onInitialize();
 }
 
-void StaticLayer::reconfigureCB()
+void
+StaticLayer::getParameters()
 {
-  RCLCPP_DEBUG(node_->get_logger(), "StaticLayer:: Event Callback");
+  int temp_lethal_threshold;
 
-  bool enabled = true;
-  dynamic_param_client_->get_event_param_or(name_ + "." + "enabled", enabled, true);
```

Comment:
Plugins are loaded during on_configure. If one wanted to load a different set of plug-ins, the node could be brought back to the initial state (deactivate, cleanup), the parameter specifying the list of plugins changed, and then configured and activated. However, if one wants to dynamically enable and disable a loaded plugin, that depends on the programming model. We just need to decide which state the enabled parameter is associated with. For example, we could require that the plugin is in the configured/inactive state, or it could be updated in the active state. In general, we should scrutinize every parameter in the system and decide a) is the parameter needed? b) in what state must the node be in for the parameter to be changed?